Build previous releases and run functional tests

Sjors opened the pull request:
This PR adds binaries for 0.17, 0.18 and 0.19 to Travis and runs a basic block propagation test.

Includes test for upgrading v0.17.1 wallets and opening master wallets with older versions.

Usage:

```sh
contrib/devtools/previous_release.sh -f -b v0.19.0.1 v0.18.1 v0.17.1
test/functional/backwards_compatibility.py
```

Travis caches these earlier releases, so it should be able to run these tests with little performance impact.

Additional scenarios where it might be useful to run tests against earlier releases:

* creating a wallet with #11403's segwit implementation, copying it to an older node and making sure the user didn't lose any funds (although this PR doesn't support `v0.15.1`)
* future consensus changes
* P2P changes (e.g. to make sure we don't accidentally ban old nodes)

ryanofsky commented:
Can you explain why exactly the patching is needed in this case? It would definitely seem better to test against unpatched previous releases if possible.

Sjors commented:
@ryanofsky the regtest parameters were changed to activate SegWit at the genesis block, which causes nodes after #11389 to reject blocks created by v0.15.1 and older nodes (and vice versa). Maybe there's a less drastic way to change the regtest consensus params for these older releases?

ryanofsky commented:
Not sure, but is passing `-vbparams=segwit:0:0` an option?

sdaftuar commented:
@sjors I've spent some time looking into how feasible it would be to write functional tests that would run against older versions of our software, which I think would be great if we could pull off -- for instance, I thought it would be nice in something like the `p2p-segwit.py` test for us to use an actual old binary in the `upgrade_after_activation` test, rather than mocking it with a current binary by setting the bip9 parameters to 0.

However, in my experience this proved to be impractical.  It seems like we periodically make changes to the RPC test framework infrastructure itself, in such ways that cause problems with backward compatibility.  Here are a few examples I have run into, so you get a flavor of some of the issues:

- `sync_blocks()` was rewritten to use an rpc call, `waitforblockheight`, that was introduced in 0.14.  So no test that uses sync_blocks can run on a pre-0.14 version of the code (which is basically every test we have!).

- Even if you solve that problem somehow (eg by rewriting `sync_blocks`, or avoiding it somehow), we also introduced in 0.14 support for RPC named arguments.  This also breaks compatibility with pre-0.14 software (so without a patch, I believe no RPC calls work at all).

- I might have the details of this one slightly wrong, but I think we made a change at some point to support parsing numbers as strings, versus floats, and then I think our JSON library started to take advantage of that as well?  Not sure if that's exactly right, but I remember having an issue running newer tests against older releases when using RPC calls involving numbers, and needing a patch for that.

So my takeaway from this exercise was that if we want to support running modern tests against older binaries, we need to change the way we do things, so that we (a) have a better split between the python test code, and the python utilities/RPC handlers/etc that allow us to talk to a node, and (b) come up with a way to use older python utility code when talking to older nodes.

ryanofsky commented:
> So no test that uses sync_blocks can run on a pre-0.14 version of the code (which is basically every test we have!).

I don't understand why this would matter. I think the goal here is just to be able to write a small test that brings up a specific version of bitcoind (0.15) and tests a few things with it, not to be able to run large swathes of the test framework against arbitrary bitcoin releases. It doesn't seem it would require a lot of code to support this, and if the tests are run on travis it doesn't seem like whatever code is added could get broken easily.

sdaftuar commented:
@ryanofsky Perhaps I phrased that sentence poorly -- my point wasn't that we'd want to run all the tests against older code (I think that would be absurd), just that we take it for granted that we can call `sync_blocks` in all sorts of places, so to not have such a basic utility at your disposal can be surprising and hard to work around.

Another way to look at it: imagine you wanted to run a test against 0.15 now, so you write one and you can even use `sync_blocks` and everything works fine (I agree that this shouldn't be too hard, as 0.15 was very recent).  Based on our history, I imagine there will come a time when someone wants to do something which will break the test, because some utility code -- like `sync_blocks`, or the way we make RPC calls, or something else -- will change, and the test will be busted, and we won't have a good way to fix it.

jnewbery commented:
This seems possible for small, limited tests which don't use the full range of test framework capabilities (for example, if we just want to test downgrading to v0.15 to test segwit wallets). As Suhas has pointed out, there are lots of reasons why this is difficult in the general case, but I think we can work around them for a targeted test cases:

> sync_blocks() was rewritten to use an rpc call, waitforblockheight, that was introduced in 0.14. 

Not relevant if we only want to test as far back as v0.15, but we can add a `waitforblockheight` method to `TestNode` for versions that don't have a `waitforblockheight` RPC method.

> Even if you solve that problem somehow (eg by rewriting sync_blocks, or avoiding it somehow), we also introduced in 0.14 support for RPC named arguments.

We could be careful to not use named arguments in tests where we want to use earlier versions. Or we could add a shim to `TestNode` to convert named -> positional.

> I think we made a change at some point to support parsing numbers as strings, versus floats, and then I think our JSON library started to take advantage of that as well?

As above.

> I imagine there will come a time when someone wants to do something which will break the test, because some utility code -- like sync_blocks, or the way we make RPC calls, or something else -- will change, and the test will be busted, and we won't have a good way to fix it.

I expect we'll only want to have tests covering the last one or two releases, so I don't see this as a huge problem.

> I think the goal here is ... not to be able to run large swathes of the test framework against arbitrary bitcoin releases.

Yes - I agree that would be almost impossible.

@Sjors - if the goal is only to test official releases, why not download the binaries directly from https://bitcoincore.org/bin/ rather than building them locally?

sdaftuar commented:
This is an aside if the goal here is just to make 0.15 work, but:

> We could be careful to not use named arguments in tests where we want to use earlier versions. Or we could add a shim to TestNode to convert named -> positional.

The change to support named args actually affected RPC invocations that don't use named args -- I think we send an empty dictionary instead of an empty list in `authproxy.py` which an old server wouldn't accept.

Anyway I don't mean to be overly pessimistic if you all think we can make something work!  I'm definitely in favor of the overall goal, I just haven't come across a simple solution that I thought was robust.

sdaftuar commented:
> I expect we'll only want to have tests covering the last one or two releases, so I don't see this as a huge problem.

I don't follow this point though -- in something like the `p2p-segwit.py` test, you'd have to do the upgrade after activation test on an 0.13.0 node or earlier.  So if we'd written the test in such a way as to use an old binary, we'd have to just decommission that test at some point...?  Obviously replacing with a later binary (that has the feature you want to test upgrade from) would mean you can't do that test anymore.

jnewbery commented:
> in something like the p2p-segwit.py test, you'd have to do the upgrade after activation test on an 0.13.0 node or earlier. So if we'd written the test in such a way as to use an old binary, we'd have to just decommission that test at some point...?

Yes - my expectation is that we'd decommission that test at some point. I don't think that testing versions older than a couple of releases is something that we should do as part of travis or developer local testing (although there's nothing stopping anyone from having a custom test rig to do that sort of thing).

Sjors commented:
@jnewbery compiling from source seems more flexible than fetching binaries, especially if we need to patch things. There might also be configure flags that are optimal for the functional tests. It can be expanded to fetch code from other repositories (on your own test rig). 

Downloading binaries would be faster, but it also involves figuring out which OS you're on and checking the checksum, so it's not necessarily easier to write a script for. That said, the one-off build is painfully slow on Travis atm, so I might add an option to just fetch a binary.

@ryanofsky `-vbparams=segwit:0:0 -vbparams=csv:0:0` did the trick, as long as I generate blocks on the modern node (which is fine). So no patch needed for `v0.15.1` at the moment.

@sdaftuar I ran into the issue you mentioned for versions before v0.14 and added a comment to the test file to warn about that.

I changed the demonstration test to simply check if old nodes sync blocks from the new node. I'll try some SegWit wallet related scenarios in a different PR.

I added a commit that makes one Travis instance compile `v0.14.2` and `v0.15.1` (cached) and then run the relevant functional test. It works, but I'm a bit confused as to what the right place is on Travis to put these older builds.

MarcoFalke commented on .travis.yml:
```diff
@@ -7,6 +7,7 @@ cache:
   - depends/built
   - depends/sdk-sources
   - $HOME/.ccache
+  - build/releases
```
That cache should be covered by `.ccache`, no?

Sjors replied:
- [ ] will look into `ccache` behavior.

MarcoFalke replied:
Forget what I said. It is probably better to cache the resulting binary than the individual obj files, since we are building specific tags that don't change. For master or other branch build, the commits often only change parts of the code, so the other obj files can be cached and re-used. For tag builds everything is static, so caching the binaries works as well.

MarcoFalke commented on test/functional/backwards_compatibility.py:
```diff
@@ -0,0 +1,48 @@
+        """Override this method to customize test node setup"""
+        self.add_nodes(self.num_nodes, self.extra_args, None, None, [
+            None,
+            "build/releases/v0.15.1/src/bitcoind",
```
Would be nice to be able to pass those paths in. Could overwrite `add_options` in test_framework.

Sjors replied:
Do you mean making `--srcdir` accept multiple paths?

MarcoFalke commented on test/functional/backwards_compatibility.py:
```diff
@@ -0,0 +1,48 @@
+        ]
+
+    def setup_nodes(self):
+        """Override this method to customize test node setup"""
```
Comment seems misplaced

Sjors replied:
Removed.

MarcoFalke commented on test/functional/backwards_compatibility.py:
```diff
@@ -0,0 +1,48 @@
+
+    def setup_nodes(self):
+        """Override this method to customize test node setup"""
+        self.add_nodes(self.num_nodes, self.extra_args, None, None, [
```
Use named args for those unnamed arguments to provide documentation and robustness against future interface changes.

Sjors replied:
Done.

maflcko reviewed (commented):
Concept ACK. Just noting that every change to the rpc interface is a breaking change in compatibility tests, that potentially needs attention. So we need to run those tests in the pull request travis run.

jonasschnelli commented:
Concept ACK, haven't looked at the code.

I think not testing with older versions is the biggest lack in our test env.
Things like the UTXO migration (now done) or testing how "older" nodes act on the (new) p2p network.
Also, up- and downgrading wallets is not covered by our tests and could be with something like this.

MeshCollider commented:
Concept ACK, cross version testing will be a very nice addition to the testing suite IMO

snuggs commented on .travis.yml:
```diff
@@ -70,7 +72,7 @@ script:
     - if [ -z "$NO_DEPENDS" ]; then depends/$HOST/native/bin/ccache --max-size=$CCACHE_SIZE; fi
     - test -n "$USE_SHELL" && eval '"$USE_SHELL" -c "./autogen.sh"' || ./autogen.sh
-    - mkdir build && cd build
+    - mkdir -p build && cd build
```
Nice nit @MarcoFalke!

Sjors commented:
@MarcoFalke wrote:
> we need to run those tests in the pull request travis run

They ran for this PR, so I assume they'll run on every PR once this is merged?

Sjors commented:
I found this quite useful in testing the recently merged SegWit wallet changes in WIP #12152.

I didn't need more changes to the build script and Travis configuration for those tests. On other hand in order to swap wallets in a test I'll need more involved changes to the test framework as well as dynamic loading of wallets. So I'll keep the tests in this PR trivial.

With the above caveats, noting this is probably only maintainable for the most recent one or two releases, I think it's ready for review.

Once everyone is happy with the code, someone with Travis privileges should trigger a new build with a completely wiped cache (I can also add and drop a commit that uses the `-r` flag to force a rebuild, but that's less rigorous).

Sjors commented:
Rebased. I added support for `v0.16.0` and removed `v0.15.1` and `v0.14.2` since they're no longer compatible with the test framework (can perhaps be patched later).

MarcoFalke commented on .travis.yml:
```diff
@@ -62,6 +63,7 @@ before_script:
     - if [ -z "$NO_DEPENDS" ]; then make $MAKEJOBS -C depends HOST=$HOST $DEP_OPTS; fi
     # Start xvfb if needed, as documented at https://docs.travis-ci.com/user/gui-and-headless-browsers/#Using-xvfb-to-Run-Tests-That-Require-a-GUI
     - if [ "$NEED_XVFB" = 1 ]; then export DISPLAY=:99.0; /sbin/start-stop-daemon --start --pidfile /tmp/custom_xvfb_99.pid --make-pidfile --background --exec /usr/bin/Xvfb -- :99 -ac; fi
+    - if [ "$COMPILE_PREVIOUS_VERSIONS" = "true" ]; then contrib/devtools/previous_release.sh -d -f v0.16.0; fi
```
Just noting that in case we don't apply local patches to the previous release, we might as well just fetch the gitian result from one of the download sites. That would also save us from having to deal with caching the binaries et al.

Sjors replied:
Maybe fetching from a download site could be an optional argument for `previous_release.sh`?

Sjors replied:
I'm worried that if we go for the binary download approach, the compile from source stuff would no longer be maintained. I could remove that entirely, but then some small future change to the test framework might force us to put it back in. I don't know how likely that is, but caching these files doesn't seem like a huge burden either.

Sjors commented:
Rebased

Sjors commented:
Travis failed after 50 minutes due to a timeout. Preventing that would be an argument in favor of just fetching binaries.

Sjors commented:
Rebased, and switched the test from v0.16.0 to v0.16.1

Getting another timeout during the build, so I'll add support to just fetch the binaries.

Sjors commented:
It now fetches the v0.16.1 binary instead of compiling, which should speed up Travis.

Unfortunately now the test framework seems unable to connect to the RPC, at least on my Mac.

Sjors committed:
[scripts] build earlier releases

Sjors commented:
Same problem on Travis: test framework can't connect to the v0.16.1 RPC for some reason. @jnewbery any recent changes which could explain that? Or perhaps there's something about the release binaries that gets in the way?

MarcoFalke commented:
@Sjors I believe you could add some prints in `wait_for_rpc_connection` to find out why it fails.

jnewbery commented:
If you can repro this locally, I'd recommend adding a breakpoint immediately before the call to `wait_for_rpc_connection()` and then seeing if you can connect to the RPC manually. Perhaps your binaries aren't picking up all the correct config and command line arguments?

Sjors commented:
Rebased. I'm just going to wait for the 0.17.0 release rather than try to debug why `v0.16.2` doesn't play nicely. Based on local testing that should work.

Sjors commented:
Rebased again. Moved to the `x86_64 Linux (no depends, only system libs)` host since that one has less to do.

practicalswift commented on contrib/devtools/previous_release.sh:
```diff
@@ -0,0 +1,134 @@
+  RELEASE_PATH="$PWD/build/releases/$tag"
+
+  if [ "$DELETE_EXISTING" -eq "1" ]; then
+    rm -rf $RELEASE_PATH
```
This seems very risky! :-\

Sjors replied:
I can leave out the `-f`.

practicalswift commented on contrib/devtools/previous_release.sh:
```diff
@@ -0,0 +1,134 @@
+        exit 1
+      fi
+
+      git clone . $RELEASE_PATH
```
Double quote to prevent globbing and word splitting. Please fix throughout in this PR :-)

jnewbery commented:
I'm not convinced that the extra machinery that this PR introduces is worth the additional test coverage that it provides. At the moment, the test only checks that a v0.17 node can sync with blocks generated on a master branch node. I'd expect an inconsistency that crass would be caught very quickly regardless of this test. On the other hand, extending the test to cover more complex scenarios will probably run into the difficulties enumerated by Suhas here: https://github.com/bitcoin/bitcoin/pull/12134#issuecomment-356416607. Any such test would also require maintenance every time there's a new release.

In general, I think that cross-version compatibility testing is definitely worthwhile, but I'm not convinced that it should live in the core repo and run in travis. What would your thoughts be about running this style of tests separately from our CI?

ryanofsky commented:
This change seems like it'd be useful in order to write tests for wallet upgrade code, where we have some issues to fix like #14422. If we aren't going to merge this, how can we write tests for code that handles upgrades or downgrades?

Sjors commented:
The main use case, which I think is worth some overhead, is testing soft forks before they're deployed. We could have the test craft more interesting transactions and blocks on master.

Some aspects of wallet upgrades can be tested without this, e.g. we can store older wallet payloads in the test folder. This PR lets you test opening wallets on a downgraded node. Node downgrades in general can be tested this way, by copying the datadir.

As long as I occasionally rebase this PR it should be useful even if it's not merged.

jnewbery commented:
> The main use case, which I think is worth some overhead, is testing soft forks before they're deployed. We could have the test craft more interesting transactions and blocks on master.

> As long as I occasionally rebase this PR it should be useful even if it's not merged.

Yes, this sounds reasonable. I don't think the current test case adds much, but if there's a fork in future, a test using this infrastructure would be very useful.

> Some aspects of wallet upgrades can be tested without this, e.g. we can store older wallet payloads in the test folder. This PR lets you test opening wallets on a downgraded node. Node downgrades in general can be tested this way, by copying the datadir.

Great - that sounds like a much better way of testing wallet upgrades.

practicalswift commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -0,0 +1,64 @@
+# Copyright (c) 2018 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Backwards compatibity functional test
```
Should be "compatibility" :-)

practicalswift commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -0,0 +1,64 @@
+
+        sync_blocks(self.nodes)
+
+        # Santity check the test framework:
```
Should be "sanity" :-)

jtimon commented:
Concept ACK, perhaps the python test could optionally take a custom value even if it "build/releases/v0.17.0/bin/" as default?

practicalswift commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -0,0 +1,64 @@
+    sync_blocks
+)
+
+class BackwardsCompatibityTest(BitcoinTestFramework):
```
Should be "BackwardsCompatibilityTest" :-)

practicalswift commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -0,0 +1,64 @@
+        res = self.nodes[self.num_nodes - 1].getblockchaininfo()
+        assert_equal(res['blocks'], 101)
+if __name__ == '__main__':
+    BackwardsCompatibityTest().main()
```
Same typo here :-)

Sjors commented:
(I'll rebase and fix above nits  after 0.17.1 final is tagged)

Sjors committed:
[tests] check v0.17.1 and v0.18.1 backwards compatibility

Sjors commented:
Rebased for v0.17.1

ef7c54c1eceb0eb938dd4e8312283c7433fe0c8f added a `wait` argument to `stop`, which causes the v0.17.1 node to fail upon shutdown. I worked around this by adding an optional `version` to `TestNode()`.

Empact commented on contrib/devtools/previous_release.sh:
```diff
@@ -0,0 +1,131 @@
+
+pushd "$PWD/build/releases" || exit 1
+
+  for tag in "$@"
```
nit: whitespace

Sjors replied:
See also this discussion: https://github.com/mvdan/sh/issues/293

What do you think of the indenting with brackets suggestion?

pushd somewhere || exit 1
{
   do something
}
popd || exit 1

practicalswift replied:
@Sjors Brackets make it clear that the indentation is intentional :-)

jnewbery commented:
@Sjors - what's your intention for this branch? Do you want it to get merged into master, or are you maintaining it as a separate branch for your own testing? (if so perhaps you could add a [do not merge] tag to the PR title)

The `version` logic that you've added to `TestNode` is fairly limited now, but I expect that if you want to maintain this branch and have it remain compatible with different versions of bitcoind, the complexity will start to creep up. I think this kind of testing is useful, particularly for any future softfork, but my personal preference would still be to not have this merged into master.

Sjors commented:
@jnewbery I would like it to get merged, but I'm willing to maintain it for the time being even if it doesn't get merged. I think people were worried about maintainability if this is merged. So far it's not too bad in my experience.

I don't think the goal is to support an infinite number of old bitcoind versions, so complexity can probably be kept low by dropping old versions and simplifying the test patches.

Sjors committed:
[tests] add wallet backwards compatility tests

Sjors committed:
[test] add v0.17.1 wallet upgrade test

Sjors commented:
Rebased and prepared for v0.18.0 release (just need to uncomment a few lines after the binaries are up, but it can be merged before that if we want).

I added tests for opening regular, private key disabled and blank wallets in v0.17.1 (cc @achow101, @instagibbs). They work by creating wallets in v0.18.0 and on master, copying and then opening with v0.18.0 and v0.17.1.

I also added a test to upgrade a v0.17.1 wallet to master, in particular to check if the key origin data is upgraded.

Sjors committed:
[scripts] support release candidates of earlier releases

Sjors commented:
It now uses v0.18.0 release.

practicalswift commented:
Concept ACK

Nice work!

instagibbs commented:
> and the test will be busted, and we won't have a good way to fix it

I think this is an argument for keeping the scope reduced but not convincing to keep out entirely. I sincerely hope that the functional suite can support semi-modern versions with respect to block sync checks, for example.

concept ACK

lack of wallet upgrade tests are some of the most painful points right now

Sjors commented:
Rebased to to support the latest v0.18.1. Added a version conditional for `-logthreadnames` (#15849).

Sjors commented:
Rebased after #16582, moved Travis machine 8, because machine 7 has `--disable-wallet`. Travis is having a hard time at the moment though; I tried a few restarts of machine 5 and 8 in vain.

Sjors commented:
Expanded wallet tests for #16624.

Sjors commented:
Moving releases to host-specific directory on Travis. Added a check to prevent the test from being silently skipped when it can't find the release binaries.

Sjors commented:
I'm confused by Travis / Docker file system organization. See e.g. this [log](https://api.travis-ci.org/v3/job/572928725/log.txt).

The releases are downloaded to `/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/releases`, which is cached. I don't know yet if the cache works, because so far my builds have failed so the cache didn't get stored. If it works, you should see "Using cached v0.18.1" and "Using cached v0.17.1" in the log.

`feature_backwards_compatibility.py` complains that it can't find `/home/travis/build/bitcoin/bitcoin/build/bitcoin-x86_64-unknown-linux-gnu/releases`. I don't understand why. Does stuff get moved around or am I accidentally downloading stuff into a Docker volume that the test runner doesn't see?

Sjors commented:
Hopefully solved by mounting the releases folder as another Docker volume (moved it to `$TRAVIS_BUILD_DIR/releases/$HOST`).

Sjors commented:
Getting closer! This time the wallet test actually ran and failed: `Wallet file verification failed: wallet.dat corrupt, salvage failed (-4)` ([log](https://travis-ci.org/bitcoin/bitcoin/jobs/573102685#L7596)). This happens when 0.18 tries to load a (blank) wallet from master.

Sjors committed:
[test] add 0.19 backwards compatibility tests

Sjors commented:
Rebased and added tests for v0.19.0 (using release candidate).

Sjors commented:
Rebased again. Now using v0.19.0.1

Sjors commented:
Rebased. Reminder to self: cherry-pick https://github.com/Sjors/bitcoin/commit/bacf55990905c0257c7bc45bbbd2b028716efc7f after #18067 is merged.

MarcoFalke commented on contrib/devtools/previous_release.sh:
```diff
@@ -0,0 +1,149 @@
+    fi
+  done
+}
+popd || exit 1
```
I don't think we have anyone interested in tests who can also review bash scripts. Maybe rewrite this in python?

Sjors replied:
Added to my todo list, but equally happy to (see someone else) do this in a followup.

MarcoFalke commented on ci/test/05_before_script.sh:
```diff
@@ -35,3 +35,8 @@ if [ -z "$NO_DEPENDS" ]; then
 fi
+if [ "$TEST_PREVIOUS_RELEASES" = "true" ]; then
+  BEGIN_FOLD previous-versions
+  DOCKER_EXEC contrib/devtools/previous_release.sh -f -b -t "$PREVIOUS_RELEASES_DIR" v0.17.1 v0.18.1 v0.19.0.1
```
I think `-b` implies that `-f` is unused. So why pass it?

Sjors replied:
It was a left-over from when this PR still compiled from source. Removed.

MarcoFalke commented on ci/test/04_install.sh:
```diff
@@ -62,6 +63,7 @@ if [ -z "$RUN_CI_ON_HOST" ]; then
                   --mount type=bind,src=$BASE_ROOT_DIR,dst=/ro_base,readonly \
                   --mount type=bind,src=$CCACHE_DIR,dst=$CCACHE_DIR \
                   --mount type=bind,src=$DEPENDS_DIR,dst=$DEPENDS_DIR \
+                  --mount type=bind,src=$PREVIOUS_RELEASES_DIR,dst=$PREVIOUS_RELEASES_DIR \
```
Not sure if it is worth it to cache this folder. No strong opinion though. If you keep this, please fixup the ci/README.md where it says "cache folder, like ..."

Sjors replied:
In my experience the release download is pretty slow. I updated the ci/README

MarcoFalke commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -0,0 +1,335 @@
+        if not os.path.isdir(releases_path):
+            if os.getenv("TEST_PREVIOUS_RELEASES") == "true":
+                raise AssertionError("TEST_PREVIOUS_RELEASES=1 but releases missing: " + releases_path)
+            raise SkipTest("This test requires binaries for previous releases")
```
I'd rather have both of these an Assertion instead of a silent pass

Sjors replied:
That seems inconsistent with the way we skip wallet tests, i.e. the test runner calls each test file and each test file decides if needs to be skipped.

MarcoFalke commented on test/functional/feature_backwards_compatibility.py:
```diff
@@ -0,0 +1,335 @@
+        node_v18.loadwallet("w1")
+        wallet = node_v18.get_wallet_rpc("w1")
+        info = wallet.getwalletinfo()
+        assert(info['private_keys_enabled'])
```
style-nit: 
```suggestion
        assert info['private_keys_enabled']
```

(Applies throughout this file)

Sjors replied:
Done

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -369,7 +369,7 @@ def run_test(self):
     # Public helper methods. These can be accessed by the subclass test scripts.
 
-    def add_nodes(self, num_nodes, extra_args=None, *, rpchost=None, binary=None):
+    def add_nodes(self, num_nodes, extra_args=None, *, rpchost=None, binary=None, binarycli=None, versions=None):
```
style-nit:

```suggestion
    def add_nodes(self, num_nodes, extra_args=None, *, rpchost=None, binary=None, binary_cli=None, versions=None):
```

Sjors replied:
Done

maflcko reviewed (approved):
ACK, some style nits

MarcoFalke commented:
ACK c456145b2c65f580683df03bf10cd39000cf24d5 🔨

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK c456145b2c65f580683df03bf10cd39000cf24d5 🔨
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiCZgwAwkfNVLMvxlvocPpi0dT2TM8hPUkhwYfpWFP2GFvo6BDwivVgIgRd+Wdm
O3xDZ+b+KbNfGotMRjZgz0zWpwjCef5kaJBdrG8Je9xZeU3gPmVDZMFDvoQkE7Lv
RnRhui33JaSr+TGiPHMEtHOra6+Y3frY77SFqeJw8I6SPUca0uhCllULq9ejT6k/
QeMGdZIWyRcghbvqEJNsOyUi+bqeAtl8N4YI78WCR+IsnX463RA/ky49aRvnCPma
7r8jIV548gb0AlE05C6xZzcU9Iqd6XVXCDEDIDxdmWQCmK2S3RCKHUWqKLOX5Z54
u58dXa/lPi79iztdiSeuZFsSBSOprq8/0OCjQy+VrMUa/DumKhAE0sLTTwdDjUke
mz6fVGUIGS5czt4QKBt2TxtPsNtfg+WGrmCXQp31h0Ds2S5+3Y8IfSpLNHXFxC9U
YnnG8FgCOvXz9OwhHqptpZVKYak3C3j76ulLWpkAXdVYTLXEXdRjJ0fOYGyeA1b+
L6l6wWU2
=vdCU
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `9f2d9c2708f034d9a99d437d02cbd47fb5e0b16b9e7cfc969396186bdeabfbbf  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401089f2d9c2708f034d9a99d437d02cbd47fb5e0b16b9e7cfc969396186bdeabfbbff0106827cbd99f62a070050597fbf8e203b208fff0107e2886209d8aa396080e6be4b91aa31008f1045e440952f0088ae57293c13dfbd20083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010bfac93e09d69da942f08ad5a0601b41308f1045e440952f0084bd5d51a551b61fc0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>
